lib/active_merchant/billing/gateways/blue_snap.rb

```diff
@@ -303,10 +307,10 @@ def add_charge_description(doc, description)
         end
       end
 
-      def add_order(doc, options)
+      def add_order(doc, options, add_3ds = false)
         doc.send('merchant-transaction-id', truncate(options[:order_id], 50)) if options[:order_id]
         doc.send('soft-descriptor', options[:soft_descriptor]) if options[:soft_descriptor]
-        add_3ds(doc, options[:three_d_secure]) if options[:three_d_secure]
+        add_3ds(doc, options[:three_d_secure]) if add_3ds && options[:three_d_secure]
```

Comment:
Do we need this `add_3ds` param here?
Can we just do this that way?
`add_3ds(doc, options[:three_d_secure]) if options[:three_d_secure]`

Reply:
yes we need it because the add_order method is used two times in the store method but we can send 3ds params only in the create_subscription method

```ruby
def store(payment_method, options = {})
        payment_method_details = PaymentMethodDetails.new(payment_method)

        MultiResponse.run do |r|
          r.process do
            commit(:store, :post, payment_method_details) do |doc|
              add_personal_info(doc, payment_method, options)
              add_fraud_info(doc, options)
              add_echeck_company(doc, payment_method) if payment_method_details.check?
              doc.send('payment-sources') do
                payment_method_details.check? ? store_echeck(doc, payment_method) : store_credit_card(doc, payment_method)
              end
              add_order(doc, options)
            end
          end

          if r.responses.last.success? && payment_method_details.card_transaction?
            options[:last_four] = r.responses.last.params["card-last-four-digits"]
            options[:card_type] = r.responses.last.params["card-type"]
            options[:vaulted_shopper_id] = r.responses.last.params["vaulted-shopper-id"]
            r.process do
              create_subscription(options)
            end
          end
        end
      end
```

Reply:
Hmm in that case, can we just move this method from `add_order` and call it inside `create_subscription` like that:

```ruby
def create_subscription(options = {})
  commit(:create_subscription, :post) do |doc|
    add_vaulted_shopper_id(doc, options[:vaulted_shopper_id])
    add_credit_card_info(doc, options)
    add_fraud_info(doc, options)
    add_order(doc, options, true)
    add_3ds(doc, options[:three_d_secure]) if options[:three_d_secure]
    doc.send('currency', options[:currency])
  end
end
```

Reply:
What do you think?